### PR TITLE
Haley Mesander 1801 Final

### PR DESCRIPTION
**Pull request by hmesander**
The final went well.  I read the first user story incorrectly at first (I was going to admin/genres instead of genres), but went back and corrected my errors.

### CODE REVIEW

File: app/controllers/admin/genres_controller.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Admin::GenresController < Admin::BaseController
+
+  def create
+    @genre = Genre.new(genre_params)
+    if @genre.save
```
Comment (iandouglas) [inline app/controllers/admin/genres_controller.rb]: nice job testing the edge cases and adding flash messages

File: app/controllers/admin/genres_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      flash[:success] = "You have created a #{@genre.name} genre."
+      redirect_to genres_path
+    else
+      flash[:error] = "Your genre was not successfully created."
+      redirect_to genres_path
```
Comment (iandouglas) [inline app/controllers/admin/genres_controller.rb]: since both the if and else are doing a redirect, you could move the redirect out of the if/else block like:
```ruby
if @genre.save
  flash[:success] = '...'
else
  flash[:error] = '...'
end
redirect_to genres_path
```

Comment (iandouglas) [reply app/controllers/admin/genres_controller.rb]: the downside to doing a redirect here instead of not namespacing this under /admin/ is that the form won't repopulate with what you have in `@genre` -- if this was a long form, users would be kinda ticked off about it.

File: db/migrate/20180417154223_create_movie_genres.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class CreateMovieGenres < ActiveRecord::Migration[5.1]
+  def change
+    create_table :movie_genres do |t|
+      t.references :movie, foreign_key: true
+      t.references :genre, foreign_key: true
+    end
```
Comment (iandouglas) [inline db/migrate/20180417154223_create_movie_genres.rb]: thanks for not putting timestamps in here!

File: spec/features/admin_can_create_new_genre_spec.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(admin)
+
+    visit genres_path
+
+    fill_in "Name", with: "SciFi"
+    click_on "Create Genre"
+
+    expect(page).to have_content("SciFi")
+  end
```
Comment (iandouglas) [inline spec/features/admin_can_create_new_genre_spec.rb]: no test for the edge case of the genre NOT saving??

File: spec/features/admin_can_create_new_genre_spec.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+describe "admin" do
+  scenario "can create a new genre" do
+    admin = User.create!(username: "admin", password: "password", role: 1)
+
+    allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(admin)
```
Comment (iandouglas) [inline spec/features/admin_can_create_new_genre_spec.rb]: nice!

File: spec/features/unregistered_user_can_see_movies_associated_with_a_genre_spec.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      movie1 = genre.movies.create!(title: "Guardians of the Galaxy", description: "Owls, maybe?", rating: 3, director_id: director.id)
+      movie2 = genre.movies.create!(title: "Suicide Squad", description: "Doesn't sound good...", rating: 2, director_id: director.id)
+
+      visit genres_path
+
+      click_on genre.name
+
+      expect(page).to have_content(movie1.title)
+      expect(page).to have_content(movie2.title)
+      expect(page).to have_content("Average Movie Rating: 2.5")
```
Comment (iandouglas) [inline spec/features/unregistered_user_can_see_movies_associated_with_a_genre_spec.rb]: I'd like to see Future Haley do something like:
```ruby
expected_average = (movie1.rating+movie2.rating)/2
...
expect(page).to have_content("Average Movie Rating: #{expected_average}")
```

File: spec/features/unregistered_user_sees_list_of_genres_spec.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+      expect(page).to have_link(genre1.name)
+      expect(page).to have_link(genre2.name)
+      expect(page).to have_link(genre3.name)
+    end
+
+    it "does not display the form to create a new genre" do
+      visit genres_path
+
+      expect(page).to_not have_content("Create Genre")
```
Comment (iandouglas) [inline spec/features/unregistered_user_sees_list_of_genres_spec.rb]: nice!

File: spec/models/movie_spec.rb [hunk 8]
```diff
@@ -3,5 +3,15 @@
...
     it {is_expected.to have_many(:actors).through(:actor_movies)}
+    it {is_expected.to have_many(:genres).through(:movie_genres)}
+  end
+  describe 'class_methods' do
+    it '.average_rating' do
+      director = Director.create!(name: "Someone!")
+      director.movies.create!(title: "Guardians of the Galaxy", description: "Owls, maybe?", rating: 3)
+      director.movies.create!(title: "Suicide Squad", description: "Doesn't sound good...", rating: 2)
+
+      expect(Movie.average_rating).to eq(2.5)
```
Comment (iandouglas) [inline spec/models/movie_spec.rb]: hooray for model testing!! 😍
